### PR TITLE
PERF-#367: Use std::fill to fill service shared buffer with a given value

### LINKED ISSUE #367: First put into the shared memory is slower than subsequent one
```python
import unidist
import numpy as np
import time

unidist.init()

t0 = time.time()
data = unidist.put(np.arange(10))
t1 = time.time()
print(f"time is {t1 - t0}")

t0 = time.time()
data2 = unidist.put(np.arange(10))
t1 = time.time()
print(f"time is {t1 - t0}")

# time is 2.5319178104400635
# time is 0.06026887893676758
```

Most of the time is spent in https://github.com/modin-project/unidist/blob/fea4507b2bd29ca209f1778de1cb8347de9ee804/unidist/core/backends/mpi/core/communication.py#L981

### PR DESCRIPTION
**Pull request by arunjose696**
## What do these changes do?

In _allocate_shared_memory(self): function we [are setting -1 to the service buffer because 0 is a valid value](https://github.com/modin-project/unidist/blob/master/unidist/core/backends/mpi/core/shared_object_store.py#L191-L195). This operation is a time consuming operation. Changing this to use numpy to speed up the step.
- [ ] first commit message and PR title follow format outlined [here](https://unidist.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 .`
- [ ] passes `black .`
- [x] signed commit with `git commit -s` 
- [ ] Resolves #367
- [ ] tests added and passing
- [ ] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: unidist/core/backends/mpi/core/monitor/loop.py [hunk 1]
```diff
@@ -170,6 +170,7 @@ def monitor_loop():
     The loop exits on special cancelation operation.
     ``unidist.core.backends.mpi.core.common.Operations`` defines a set of supported operations.
     """
+    monitor_logger.debug("Monitor loop started")
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/monitor/loop.py:173]:
```suggestion
```

File: unidist/core/backends/mpi/core/monitor/loop.py [hunk 2]
```diff
@@ -182,7 +183,8 @@ def monitor_loop():
     # Once all workers excluding ``Root`` and ``Monitor`` ranks are ready to shutdown,
     # ``Monitor` sends the shutdown signal to every worker, as well as notifies ``Root`` that
     # it can exit the program.
-
+    # Barrier in monitor process to check if monitor loop has started
+    mpi_state.comm.Barrier()
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/monitor/loop.py:187]:
```suggestion
    mpi_state.comm.Barrier()
    monitor_logger.debug("Monitor loop started")
```

File: unidist/core/backends/mpi/core/controller/api.py [hunk 3]
```diff
@@ -261,6 +261,8 @@ def init():
         atexit.register(_termination_handler)
         signal.signal(signal.SIGTERM, _termination_handler)
         signal.signal(signal.SIGINT, _termination_handler)
+        # Exit the init function in root only after monitor and worker loops have started.
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:264]:
```suggestion
        # Exit the init function in root only after monitor and worker loops have started
```

File: unidist/core/backends/mpi/core/monitor/loop.py [hunk 4]
```diff
@@ -182,7 +183,8 @@ def monitor_loop():
     # Once all workers excluding ``Root`` and ``Monitor`` ranks are ready to shutdown,
     # ``Monitor` sends the shutdown signal to every worker, as well as notifies ``Root`` that
     # it can exit the program.
-
+    # Barrier in monitor process to check if monitor loop has started
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/monitor/loop.py:186]:
```suggestion
    # Barrier to check if monitor process is ready to start the communication loop
```

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 5]
```diff
@@ -190,9 +190,8 @@ def _allocate_shared_memory(self):
         self.service_shared_buffer = memoryview(service_buffer).cast("l")
         # Set -1 to the service buffer because 0 is a valid value and may be recognized by mistake.
         if mpi_state.is_monitor_process():
-            self.service_shared_buffer[:] = array(
-                "l", [-1] * len(self.service_shared_buffer)
-            )
+            memoryview_array = np.frombuffer(self.service_shared_buffer, dtype=np.int64)
+            memoryview_array[:] = -1
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/shared_object_store.py:194]:
```suggestion
            values_to_fill = array("l", [-1] * len(self.service_shared_buffer))
            values_to_fill_view = memoryview(values_to_fill)
            parallel_memcopy(
                values_to_fill_view,
                self.self.service_shared_buffer[:len(self.service_shared_buffer)],
                6,
            )
```
To not introduce a new dependency (numpy) let's try to use our own function to set the value.

File: unidist/core/backends/mpi/core/worker/loop.py [hunk 6]
```diff
@@ -84,6 +84,7 @@ async def worker_loop():
     The loop exits on special cancelation operation.
     ``unidist.core.backends.mpi.core.common.Operations`` defines a set of supported operations.
     """
+    w_logger.debug("Worker loop started")
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/worker/loop.py:87]:
```suggestion
```

File: unidist/core/backends/mpi/core/worker/loop.py [hunk 7]
```diff
@@ -94,6 +95,9 @@ async def worker_loop():
     # and the worker sends the ready to shutdown signal to ``Monitor``.
     # Once all workers excluding ``Root`` and ``Monitor`` ranks are ready to shutdown,
     # ``Monitor` sends the shutdown signal to every worker so they can exit the loop.
+
+    # Barrier in worker process to check if worker loop has started
+    mpi_state.comm.Barrier()
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/worker/loop.py:100]:
```suggestion
    mpi_state.comm.Barrier()
    w_logger.debug("Worker loop started")
```

File: unidist/core/backends/mpi/core/worker/loop.py [hunk 8]
```diff
@@ -94,6 +95,9 @@ async def worker_loop():
     # and the worker sends the ready to shutdown signal to ``Monitor``.
     # Once all workers excluding ``Root`` and ``Monitor`` ranks are ready to shutdown,
     # ``Monitor` sends the shutdown signal to every worker so they can exit the loop.
+
+    # Barrier in worker process to check if worker loop has started
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/worker/loop.py:99]:
```suggestion
    # Barrier to check if worker process is ready to start the communication loop
```

File: unidist/core/backends/mpi/core/shared_object_store.py (hunk 5 above)
Comment (YarShev) [reply unidist/core/backends/mpi/core/shared_object_store.py:194]: This is as slow as the previous approach. Let's discuss offline what we should do.

File: unidist/core/backends/mpi/core/monitor/loop.py (hunk 1 above)
Comment (arunjose696) [reply unidist/core/backends/mpi/core/monitor/loop.py:173]: Added this looging line here so the init() function in root ends only after the monitor.log file is created and first log is written,  Or this file creation step though not very time consuming happens after init() of the root process and could make the first put() slightly slower as the monitor will be ready to recv operations only after creating the logger file.

Comment (YarShev) [reply unidist/core/backends/mpi/core/monitor/loop.py:173]: Discussed offline and decided to apply suggestion.

Aside: Is the log file created on the first write or when we call get_logger?

File: unidist/core/backends/mpi/core/memory/memory.h [hunk 9]
```diff
@@ -17,6 +17,8 @@ namespace unidist {
                         int64_t nbytes,
                         uintptr_t block_size,
                         int num_threads);
+    
+    void fill(int64_t *buff,int64_t size, int64_t value);
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/memory/memory.h:21]:
```suggestion
    void fill(int64_t *buff, int64_t size, int64_t value);
```

**Comment by YarShev** [comment]
Please also rename the PR title and exclude the file `unidist/core/backends/mpi/core/memory/_memory.cpp` (it is auto-generated by cython) out of this PR.

File: unidist/core/backends/mpi/core/monitor/loop.py (hunk 1 above)
Comment (YarShev) [reply unidist/core/backends/mpi/core/monitor/loop.py]: ping

**Comment by YarShev** [comment]
Some perf measurements would also be helpful.

**Review by YarShev — APPROVED** [review]

File: unidist/core/backends/mpi/core/monitor/loop.py (hunk 1 above)
Comment (arunjose696) [reply unidist/core/backends/mpi/core/monitor/loop.py]: Verified again the log file is created when the "first write" is made.

Comment (YarShev) [reply unidist/core/backends/mpi/core/monitor/loop.py]: Thanks!

**Pull request merged**